url5583f3dbb2_XOR19_XScript_tgz-pJ8-xscript_interactive_MainJ8/src/xscript/compiler/inst/XInstSwitch.java

```diff
@@ -0,0 +1,276 @@
+		}
+	}
+
+	@Override
+	public boolean isNormalJump() {
+		return false;
+	}
+
+	@Override
+	public boolean isAlwaysJump() {
+		return true;
+	}
+
+	@Override
+	public int tryWay(XCodeGen codeGen, int programPointer, int stackSize, int[] sizes) {
+		for(Integer i:resolvedSwitches2.values()){
+			codeGen.tryWay(i, stackSize, sizes);
+		}
+		return resolvedDefault2;
+	}
+	
+	@Override
+	public void toCode(XDataOutput dataOutput) {
+		super.toCode(dataOutput);
+		dataOutput.writeShort(resolvedDefault);
+		Integer i = resolvedSwitches.get(null);
+		if(i==null){
+			dataOutput.writeShort(resolvedDefault);
+		}else{
+			dataOutput.writeShort(i);
+		}
+		i = resolvedSwitches.get(true);
+		if(i==null){
+			dataOutput.writeShort(resolvedDefault);
+		}else{
+			dataOutput.writeShort(i);
+		}
+		i = resolvedSwitches.get(false);
+		if(i==null){
+			dataOutput.writeShort(resolvedDefault);
+		}else{
+			dataOutput.writeShort(i);
+		}
+		List<Long> longSwitches = new LinkedList<Long>();
+		List<Number> doubleSwitches = new LinkedList<Number>();
+		List<String> stringSwitches = new LinkedList<String>();
+		long min = Long.MAX_VALUE;
+		long max = Long.MIN_VALUE;
+		for(Entry<Object, Integer> e:resolvedSwitches.entrySet()){
+			Object key = e.getKey();
+			if(key instanceof Long){
+				long l = ((Long)key).longValue();
+				if(l>max)
+					max = l;
+				if(l<min)
+					min = l;
+				longSwitches.add((Long)key);
+			}else if(key instanceof Float || key instanceof Double){
+				doubleSwitches.add((Number)key);
+			}else if(key instanceof String){
+				stringSwitches.add((String)key);
+			}else{
+				throw new AssertionError();
+			}
+		}
+		
+		Number[] doubleKeys = doubleSwitches.toArray(new Number[doubleSwitches.size()]);
+		Arrays.sort(doubleKeys, DOUBLE_COMPARATOR);
+		dataOutput.writeShort(doubleKeys.length);
+		for(Number key:doubleKeys){
+			if(key instanceof Float){
+				dataOutput.writeByte(0);
+				dataOutput.writeFloat((Float)key);
+			}else if(key instanceof Double){
+				dataOutput.writeByte(1);
+				dataOutput.writeDouble((Float)key);
+			}
+			dataOutput.writeShort(resolvedSwitches.get(key));
+		}
+		
+		String[] stringKeys = stringSwitches.toArray(new String[stringSwitches.size()]);
+		Arrays.sort(stringKeys);
+		dataOutput.writeShort(stringKeys.length);
+		for(String key:stringKeys){
+			dataOutput.writeUTF(key);
+			dataOutput.writeShort(resolvedSwitches.get(key));
+		}
+
+		if(longSwitches.size()==0){
+			dataOutput.writeByte(5);
+		}else if(useTabel(min, max, longSwitches.size())){
+			if(min>=Byte.MIN_VALUE && min<=Byte.MAX_VALUE){
+				dataOutput.writeByte(0);
+				dataOutput.writeByte((int)min);
+			}else if(min>=Short.MIN_VALUE && min<=Short.MAX_VALUE){
+				dataOutput.writeByte(1);
+				dataOutput.writeShort((int)min);
+			}else if(min>=Integer.MIN_VALUE && min<=Integer.MAX_VALUE){
+				dataOutput.writeByte(2);
+				dataOutput.writeInt((int)min);
+			}else{
+				dataOutput.writeByte(3);
+				dataOutput.writeLong(min);
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 158:167, 185:194.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.